File: java/Greedy/Huffman_Coding.java
package java.Greedy;

import java.util.*;

class Node {
  int data;
  char ch;

  Node left;
  Node right;

  Node(char ch, int data) {
    this.ch = ch;
    this.data = data;
  }
}

public class Huffman_Coding {
  void printCodes(Node root, String code) {
    if (root == null)
      return;

    if (root.left == null && root.right == null) {
      System.out.println(root.ch + " : " + code);
    }
    printCodes(root.left, code + "0");
    printCodes(root.right, code + "1");
  }

  public void huffman(char[] chars, int[] freq, int n) {
    PriorityQueue<Node> pq = new PriorityQueue<>((a, b) -> a.data - b.data);
    for (int i = 0; i < n; i++) {
      pq.offer(new Node(chars[i], freq[i]));
    }
    while (pq.size() > 1) {
      Node left = pq.poll();
      Node right = pq.poll();
      Node merged = new Node('-', left.data + right.data);
      merged.left = left;
      merged.right = right;
      pq.offer(merged);
    }
    printCodes(pq.peek(), "");
  }
}

File: python/Greedy/Huffman_Coding.py
import heapq 

class Node:
  def __init__(self, freq, char=None):
    self.freq = freq 
    self.char= char 
    self.left = None 
    self.right = None 

  def __lt__(self, other):
    return self.freq < other.freq 

def huffmanCoding(chars, freq):
  heap = []

  for c, f in zip(chars, freq):
    heapq.heappush(heap, Node(f, c))
  
  while len(heap) > 1:
    left = heapq.heappop(heap)
    right = heapq.heappop(heap)

    parent = Node(left.freq + right.freq)
    parent.left = left 
    parent.right = right 

    heapq.heappush(heap, parent)
  
  root = heap[0]

  codes = {}

  def dfs(node, code):
    if not node:
      return
    
    if node.char is not None:
      codes[node.char] = code 
    
    dfs(node.left, code + "0")
    dfs(node.right, code + "1")

  dfs(root, "")

  return codes 


